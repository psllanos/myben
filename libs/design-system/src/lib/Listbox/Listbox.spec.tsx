/* eslint-disable @typescript-eslint/no-empty-function */
import { fireEvent, render, screen } from '@testing-library/react'
import { Listbox } from './'

// Not tested too thoroughly, most logic is covered by Headless UI base
describe('Listbox', () => {
    describe('when closed', () => {
        it('should render the button', () => {
            const component = render(
                <Listbox value="Option 1" onChange={() => {}}>
                    <Listbox.Button>Select...</Listbox.Button>
                    <Listbox.Options>
                        <Listbox.Option value="Option 1">Option 1</Listbox.Option>
                    </Listbox.Options>
                </Listbox>
            )

            expect(screen.getByText('Select...')).toBeVisible()
            expect(screen.queryByText('Option 1')).not.toBeVisible()
            expect(component).toMatchSnapshot()
        })
    })

    describe('when opened', () => {
        it('should render the button and the items', () => {
            const component = render(
                <Listbox value="Option 1" onChange={() => {}}>
                    <Listbox.Button>Select...</Listbox.Button>
                    <Listbox.Options>
                        <Listbox.Option value="Option 1">Option 1</Listbox.Option>
                        <Listbox.Option value="Option 2" disabled={true}>
                            Option 2
                        </Listbox.Option>
                    </Listbox.Options>
                </Listbox>
            )

            fireEvent.click(screen.getByText('Select...'))
            ;['Select...', 'Option 1', 'Option 2'].forEach((text) =>
                expect(screen.getByText(text)).toBeVisible()
            )
            expect(component).toMatchSnapshot()
        })
    })

    describe('when an option is pressed', () => {
        it('should call the `onChange` callback', () => {
            const onChangeMock = jest.fn()
            render(
                <Listbox value="Option 1" onChange={onChangeMock}>
                    <Listbox.Button>Select...</Listbox.Button>
                    <Listbox.Options>
                        <Listbox.Option value="Option 1">Option 1</Listbox.Option>
                    </Listbox.Options>
                </Listbox>
            )

            fireEvent.click(screen.getByText('Select...'))
            fireEvent.click(screen.getByText('Option 1'))
            expect(onChangeMock).toHaveBeenCalledWith('Option 1')
        })
    })
})
