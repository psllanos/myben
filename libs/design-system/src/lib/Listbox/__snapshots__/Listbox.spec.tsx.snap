// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox when closed should render the button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        data-headlessui-state=""
      >
        <label
          aria-controls="headlessui-listbox-options-:r1:"
          aria-expanded="false"
          aria-haspopup="true"
          class="relative flex w-full flex-col"
          data-headlessui-state=""
          id="headlessui-listbox-button-:r0:"
        >
          <button
            class="text-white bg-gray-500 border-gray-500 focus:border-cyan focus:ring-opacity-10 focus:ring-cyan py-2 px-4 h-10 flex items-center grow rounded border overflow-hidden text-base text-left font-light leading-none focus:ring placeholder-gray-200 disabled:placeholder-gray-200"
            type="button"
          >
            <span
              class="grow"
            >
              Select...
            </span>
            <svg
              class="ml-2 text-lg text-gray-50"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 16l-6-6h12z"
                />
              </g>
            </svg>
          </button>
        </label>
        <div
          class="z-20 absolute min-w-full shadow-md rounded bg-gray-700"
          style="position: absolute; left: 0px; top: 0px;"
        >
          <ul
            aria-activedescendant="headlessui-listbox-option-:r2:"
            aria-labelledby="headlessui-listbox-button-:r0:"
            aria-orientation="vertical"
            class="py-2"
            data-headlessui-state=""
            hidden=""
            id="headlessui-listbox-options-:r1:"
            role="listbox"
            style="display: none;"
            tabindex="0"
          >
            <li
              aria-selected="true"
              data-headlessui-state="active selected"
              id="headlessui-listbox-option-:r2:"
              role="option"
              tabindex="-1"
            >
              <button
                class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap bg-gray-500 text-white"
                type="button"
              >
                <div
                  class="flex gap-x-2 items-center"
                >
                  <span>
                    Option 1
                  </span>
                </div>
                <span
                  class="text-white"
                >
                  <svg
                    class="w-5 h-5"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414z"
                      />
                    </g>
                  </svg>
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      data-headlessui-state=""
    >
      <label
        aria-controls="headlessui-listbox-options-:r1:"
        aria-expanded="false"
        aria-haspopup="true"
        class="relative flex w-full flex-col"
        data-headlessui-state=""
        id="headlessui-listbox-button-:r0:"
      >
        <button
          class="text-white bg-gray-500 border-gray-500 focus:border-cyan focus:ring-opacity-10 focus:ring-cyan py-2 px-4 h-10 flex items-center grow rounded border overflow-hidden text-base text-left font-light leading-none focus:ring placeholder-gray-200 disabled:placeholder-gray-200"
          type="button"
        >
          <span
            class="grow"
          >
            Select...
          </span>
          <svg
            class="ml-2 text-lg text-gray-50"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 16l-6-6h12z"
              />
            </g>
          </svg>
        </button>
      </label>
      <div
        class="z-20 absolute min-w-full shadow-md rounded bg-gray-700"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <ul
          aria-activedescendant="headlessui-listbox-option-:r2:"
          aria-labelledby="headlessui-listbox-button-:r0:"
          aria-orientation="vertical"
          class="py-2"
          data-headlessui-state=""
          hidden=""
          id="headlessui-listbox-options-:r1:"
          role="listbox"
          style="display: none;"
          tabindex="0"
        >
          <li
            aria-selected="true"
            data-headlessui-state="active selected"
            id="headlessui-listbox-option-:r2:"
            role="option"
            tabindex="-1"
          >
            <button
              class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap bg-gray-500 text-white"
              type="button"
            >
              <div
                class="flex gap-x-2 items-center"
              >
                <span>
                  Option 1
                </span>
              </div>
              <span
                class="text-white"
              >
                <svg
                  class="w-5 h-5"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414z"
                    />
                  </g>
                </svg>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Listbox when opened should render the button and the items 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        data-headlessui-state="open"
      >
        <label
          aria-controls="headlessui-listbox-options-:r4:"
          aria-expanded="true"
          aria-haspopup="true"
          class="relative flex w-full flex-col"
          data-headlessui-state="open"
          id="headlessui-listbox-button-:r3:"
        >
          <button
            class="text-white bg-gray-500 border-gray-500 focus:border-cyan focus:ring-opacity-10 focus:ring-cyan py-2 px-4 h-10 flex items-center grow rounded border overflow-hidden text-base text-left font-light leading-none focus:ring placeholder-gray-200 disabled:placeholder-gray-200"
            type="button"
          >
            <span
              class="grow"
            >
              Select...
            </span>
            <svg
              class="ml-2 text-lg text-gray-50"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 16l-6-6h12z"
                />
              </g>
            </svg>
          </button>
        </label>
        <div
          class="z-20 absolute min-w-full shadow-md rounded bg-gray-700"
          style="position: absolute; left: 0px; top: 0px;"
        >
          <ul
            aria-activedescendant="headlessui-listbox-option-:r5:"
            aria-labelledby="headlessui-listbox-button-:r3:"
            aria-orientation="vertical"
            class="py-2"
            data-headlessui-state="open"
            id="headlessui-listbox-options-:r4:"
            role="listbox"
            style=""
            tabindex="0"
          >
            <li
              aria-selected="true"
              data-headlessui-state="active selected"
              id="headlessui-listbox-option-:r5:"
              role="option"
              tabindex="-1"
            >
              <button
                class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap bg-gray-500 text-white"
                type="button"
              >
                <div
                  class="flex gap-x-2 items-center"
                >
                  <span>
                    Option 1
                  </span>
                </div>
                <span
                  class="text-white"
                >
                  <svg
                    class="w-5 h-5"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414z"
                      />
                    </g>
                  </svg>
                </span>
              </button>
            </li>
            <li
              aria-disabled="true"
              aria-selected="false"
              data-headlessui-state="disabled"
              id="headlessui-listbox-option-:r6:"
              role="option"
            >
              <button
                class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap text-gray-25 hover:bg-gray-600 opacity-50"
                type="button"
              >
                <div
                  class="flex gap-x-2 items-center"
                >
                  <span>
                    Option 2
                  </span>
                </div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      data-headlessui-state="open"
    >
      <label
        aria-controls="headlessui-listbox-options-:r4:"
        aria-expanded="true"
        aria-haspopup="true"
        class="relative flex w-full flex-col"
        data-headlessui-state="open"
        id="headlessui-listbox-button-:r3:"
      >
        <button
          class="text-white bg-gray-500 border-gray-500 focus:border-cyan focus:ring-opacity-10 focus:ring-cyan py-2 px-4 h-10 flex items-center grow rounded border overflow-hidden text-base text-left font-light leading-none focus:ring placeholder-gray-200 disabled:placeholder-gray-200"
          type="button"
        >
          <span
            class="grow"
          >
            Select...
          </span>
          <svg
            class="ml-2 text-lg text-gray-50"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 16l-6-6h12z"
              />
            </g>
          </svg>
        </button>
      </label>
      <div
        class="z-20 absolute min-w-full shadow-md rounded bg-gray-700"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <ul
          aria-activedescendant="headlessui-listbox-option-:r5:"
          aria-labelledby="headlessui-listbox-button-:r3:"
          aria-orientation="vertical"
          class="py-2"
          data-headlessui-state="open"
          id="headlessui-listbox-options-:r4:"
          role="listbox"
          style=""
          tabindex="0"
        >
          <li
            aria-selected="true"
            data-headlessui-state="active selected"
            id="headlessui-listbox-option-:r5:"
            role="option"
            tabindex="-1"
          >
            <button
              class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap bg-gray-500 text-white"
              type="button"
            >
              <div
                class="flex gap-x-2 items-center"
              >
                <span>
                  Option 1
                </span>
              </div>
              <span
                class="text-white"
              >
                <svg
                  class="w-5 h-5"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414z"
                    />
                  </g>
                </svg>
              </span>
            </button>
          </li>
          <li
            aria-disabled="true"
            aria-selected="false"
            data-headlessui-state="disabled"
            id="headlessui-listbox-option-:r6:"
            role="option"
          >
            <button
              class="group flex justify-between items-center px-2.5 gap-x-2 w-full text-base leading-8 whitespace-nowrap text-gray-25 hover:bg-gray-600 opacity-50"
              type="button"
            >
              <div
                class="flex gap-x-2 items-center"
              >
                <span>
                  Option 2
                </span>
              </div>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
